driver-core/src/main/com/mongodb/internal/time/Timeout.java

```diff
@@ -29,148 +31,120 @@
-     * <p>
-     * Returns {@code currentNanos} - {@link #startNanos}:
-     * <ul>
-     *     <li>
-     *         A negative value means either of the following
-     *         <ol>
-     *             <li>the clock from which {@code currentNanos} was read jumped backwards,
-     *             in which case the behaviour of this class is undefined;</li>
-     *             <li>(n * 2<sup>63</sup> - 1; (n + 1) * 2<sup>63</sup>)<sup>(*)</sup> nanoseconds has elapsed,
-     *             in which case the timeout has expired.</li>
-     *         </ol>
-     *     </li>
-     *     <li>
-     *         0 means either of the following
-     *         <ol>
-     *             <li>0 nanoseconds has elapsed;</li>
-     *             <li>(n + 1) * 2<sup>63</sup><sup>(*)</sup> nanoseconds has elapsed,
-     *             in which case the timeout has expired.</li>
-     *         </ol>
-     *         Since it is impossible to differentiate the former from the latter, and the former is much more likely to happen in practice,
-     *         this class interprets 0 value as 0 elapsed nanoseconds.
-     *     </li>
-     *     <li>
-     *         A positive value means either of the following
-     *         <ol>
-     *             <li>this exact number of nanoseconds has elapsed;</li>
-     *             <li>((n + 1) * 2<sup>63</sup>; (n + 2) * 2<sup>63</sup> - 1]<sup>(*)</sup> nanoseconds has elapsed,
-     *             in which case the timeout has expired.</li>
-     *         </ol>
-     *         Since it is impossible to differentiate the former from the latter, and the former is much more likely to happen in practice,
-     *         this class interprets a positive value as the exact number of elapsed nanoseconds.
-     *     </li>
-     * </ul>
-     * <hr>
-     * <sup>(*)</sup> n is positive and odd.
+     * @see #started(long, TimePoint)
      */
-    private long elapsedNanos(final long currentNanos) {
-        assertFalse(isInfinite() || isImmediate());
-        return currentNanos - startNanos;
+    public static Timeout infinite() {
+        return INFINITE;
+    }
+
+    /**
+     * @see #started(long, TimePoint)
+     */
+    public static Timeout immediate() {
+        return IMMEDIATE;
     }
 
     /**
      * Returns 0 or a positive value.
      * 0 means that the timeout has expired.
-     * <p>
-     * Must not be called on {@linkplain #isInfinite() infinite} timeouts.
+     *
+     * @throws AssertionError If the timeout is {@linkplain #isInfinite() infinite} or {@linkplain #isImmediate() immediate}.
      */
     @VisibleForTesting(otherwise = PRIVATE)
-    long remainingNanos(final long currentNanos) {
-        assertFalse(isInfinite() || isImmediate());
-        long elapsedNanos = elapsedNanos(currentNanos);
-        return elapsedNanos < 0 ? 0 : Math.max(0, durationNanos - elapsedNanos);
+    long saturatingRemainingNanos(final TimePoint now) {
```

Comment:
[optional] The term "saturating" was confusing for me. I had to look it up, and I think it refers to [this](https://en.wikipedia.org/wiki/Saturation_arithmetic). I think the implication is that it is "clamped" to 0, but I think this is already implied by the term "remaining".

(But, if a term should be used, I suspect that it should also be used for the `remaining` method below?)

Reply:
> this is already implied by the term "remaining"

Thanks, that was tautology. I removed "saturating".